File: Core/PythiaCore/Sources/PythiaCore/WebDAVSyncSchedule.swift
public enum PythiaWebDAVSyncUnit: String, CaseIterable {
    case minute
    case hour
    case day
    case week

    public var seconds: Int {
        switch self {
        case .minute: return 60
        case .hour: return 3_600
        case .day: return 86_400
        case .week: return 604_800
        }
    }
}

public struct PythiaWebDAVSyncSchedule: Equatable {
    public static let maximumSeconds = 366 * 86_400

    public let value: Int
    public let unit: PythiaWebDAVSyncUnit

    public init?(value: Int, unit: PythiaWebDAVSyncUnit) {
        guard value > 0, value <= Self.maximumSeconds / unit.seconds else { return nil }
        self.value = value
        self.unit = unit
    }

    public var seconds: Int { value * unit.seconds }
    public var legacyMinutes: Int { seconds / 60 }

    public static func fromLegacyMinutes(_ rawMinutes: Int) -> PythiaWebDAVSyncSchedule {
        let minutes = min(max(1, rawMinutes), maximumSeconds / 60)
        if minutes.isMultiple(of: 10_080) {
            return PythiaWebDAVSyncSchedule(value: minutes / 10_080, unit: .week)!
        }
        if minutes.isMultiple(of: 1_440) {
            return PythiaWebDAVSyncSchedule(value: minutes / 1_440, unit: .day)!
        }
        if minutes.isMultiple(of: 60) {
            return PythiaWebDAVSyncSchedule(value: minutes / 60, unit: .hour)!
        }
        return PythiaWebDAVSyncSchedule(
            value: minutes,
            unit: .minute
        )!
    }
}

File: Windows/Pythia.Windows/lib/core/webdav_sync_schedule.dart
enum WebDavSyncIntervalUnit {
  minute('minute', '分钟', 60),
  hour('hour', '小时', 3600),
  day('day', '天', 86400),
  week('week', '周', 604800);

  final String storageValue;
  final String label;
  final int seconds;

  const WebDavSyncIntervalUnit(this.storageValue, this.label, this.seconds);

  static WebDavSyncIntervalUnit fromStorage(String? raw) {
    return values.firstWhere(
      (unit) => unit.storageValue == raw,
      orElse: () => WebDavSyncIntervalUnit.hour,
    );
  }
}

class WebDavSyncSchedule {
  static const maxSeconds = 366 * 24 * 60 * 60;

  final int value;
  final WebDavSyncIntervalUnit unit;

  const WebDavSyncSchedule(this.value, this.unit);

  int get seconds => value * unit.seconds;
  int get legacyMinutes => seconds ~/ 60;

  WebDavSyncSchedule validated() {
    if (value <= 0) {
      throw const FormatException('自动同步间隔必须大于 0');
    }
    if (seconds > maxSeconds) {
      throw const FormatException('自动同步间隔不能超过 366 天');
    }
    return this;
  }

  static WebDavSyncSchedule fromLegacyMinutes(int minutes) {
    final safeMinutes = minutes <= 0 ? 60 : minutes;
    if (safeMinutes % 10080 == 0) {
      return WebDavSyncSchedule(
        safeMinutes ~/ 10080,
        WebDavSyncIntervalUnit.week,
      );
    }
    if (safeMinutes % 1440 == 0) {
      return WebDavSyncSchedule(
        safeMinutes ~/ 1440,
        WebDavSyncIntervalUnit.day,
      );
    }
    if (safeMinutes % 60 == 0) {
      return WebDavSyncSchedule(
        safeMinutes ~/ 60,
        WebDavSyncIntervalUnit.hour,
      );
    }
    return WebDavSyncSchedule(safeMinutes, WebDavSyncIntervalUnit.minute);
  }
}
